File: solutions/haskell/anagram/1/src/Anagram.hs
module Anagram (anagramsFor) where

import Data.List (sort)
import Data.Char (toUpper)

anagramsFor :: String -> [String] -> [String]
anagramsFor xs xss = filter (isAnagram xs) xss
  where
   isAnagram xs ys = (upper xs) /= (upper ys) && (eqIgnoreCase ys) 
   eqIgnoreCase ys = uxs == (unified ys)
   unified xs = sort $ upper xs
   upper = map toUpper
   uxs = unified xs
   upxs = upper xs

File: solutions/erlang/anagram/3/src/anagram.erl
-module(anagram).
-export([find_anagrams/2]).
-import(string, [to_upper/1]).

find_anagrams(S, Cs) ->
  [C || C <- Cs, is_anagram(to_upper(S), to_upper(C)) ].

is_anagram(S, S) -> false;
is_anagram(S, C) -> lists:sort(S) == lists:sort(C).
    
